### DISCUSSION TITLE
Crazyflie client on Raspberry Pi

### DISCUSSION THREAD

**Discussion by ryanerath** [bitcraze/discussions#1166] [Q&A]
Is there any guidance or documentation to install the Crazyflie client on Raspberry Pi? I tried the Linux install on the latest version of the Raspberry OS but was running into issues with PyQt. 

Thank you!

**Comment by knmcguire** [comment]
Hi! 

I'm actually not surprised that pyqt is giving issues... we haven't tried the cfclient on the latest Raspbian on a raspberry pi. and haven't for a long time. There are differences between the raspbian and an full ubuntu install.

I'm traveling myself but luckily I do have a latest raspberry 5 with me so I can take a look what  the issues is. I'll also ask @ataffanel if he knows anything about this as well.

**Comment by ataffanel** [comment]
I am also traveling and not able to test right now but can you tell me what error you are getting exactly? Testing in a debian bookworm distrobox container on my PC I am indeed getting a couple of Qt related issue that I am debugging, but they might be related to my container environment more than to Qt.

**Reply by ataffanel** [reply]
The way to get it to work on a clean distrobox debian (with Wayland which should be the same as on a pie) is:

``` bash
$ sudo apt install pipx libwayland-\* libxkbcommon0
$ pipx install cfclient
$ cfclient    # Works!
```

Installing all `libwayland` package is very likely overkil but it seems to work ... Please tell us if that helps and we can update the documentation.

**Reply by ryanerath** [reply]
Hello, 

Thank you for the response. I am still having some issues installing client with the above. Looks to still be related to pyqt. See the screenshot below. 

Thank you!

![cf](https://github.com/bitcraze/discussions/assets/110474861/da7fa1ff-7d28-4f9d-a43b-36f9f9befcfc)

**Reply by steverpalmer** [reply]
I think I'm having the same problem.  I'm using a fresh Raspberry Pi OS (64 bit) on a Pi 5.  After an `apt update && apt upgrade` && reboot, I did the `sudo apt install git python3-pip libxcb-xinerama0 libxcb-cursor0` (of which only libxcb-cursor0 is newly installed).  I also did the `sudo apt install pipx libwayland-\* libxkbcommon0`, and installed a bunch of other stuff from the QT6 package dependency list.

After `pip install --upgrade pip`, when I tried `pip install cfclient` from a virtual environment, I get:

```
> Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
> Collecting cfclient
>   Using cached https://www.piwheels.org/simple/cfclient/cfclient-2023.11-py3-none-any.whl (295 kB)
> Collecting cflib>=0.1.24 (from cfclient)
>   Using cached https://www.piwheels.org/simple/cflib/cflib-0.1.24-py2.py3-none-any.whl (189 kB)
> Collecting appdirs~=1.4.0 (from cfclient)
>   Using cached https://www.piwheels.org/simple/appdirs/appdirs-1.4.4-py2.py3-none-any.whl (9.6 kB)
> Collecting pyzmq~=25.0 (from cfclient)
>   Using cached pyzmq-25.1.2-cp311-cp311-manylinux_2_17_aarch64.manylinux2014_aarch64.whl.metadata (4.9 kB)
> Collecting pyqtgraph~=0.13 (from cfclient)
>   Using cached https://www.piwheels.org/simple/pyqtgraph/pyqtgraph-0.13.3-py3-none-any.whl (960 kB)
> Collecting PyYAML~=6.0.1 (from cfclient)
>   Using cached PyYAML-6.0.1-cp311-cp311-manylinux_2_17_aarch64.manylinux2014_aarch64.whl.metadata (2.1 kB)
> Collecting numpy<1.25,>=1.20 (from cfclient)
>   Using cached numpy-1.24.4-cp311-cp311-manylinux_2_17_aarch64.manylinux2014_aarch64.whl.metadata (5.6 kB)
> Collecting vispy~=0.13.0 (from cfclient)
>   Using cached vispy-0.13.0-cp311-cp311-manylinux_2_17_aarch64.manylinux2014_aarch64.whl.metadata (8.3 kB)
> Collecting pyserial~=3.5 (from cfclient)
>   Using cached https://www.piwheels.org/simple/pyserial/pyserial-3.5-py2.py3-none-any.whl (90 kB)
> Collecting pyqt6~=6.5 (from cfclient)
>   Using cached PyQt6-6.6.1.tar.gz (1.0 MB)
>   Installing build dependencies ... done
>   Getting requirements to build wheel ... done
>   Preparing metadata (pyproject.toml) ... error
>   error: subprocess-exited-with-error
>   
>   × Preparing metadata (pyproject.toml) did not run successfully.
>   │ exit code: 1
>   ╰─> [26 lines of output]
>       pyproject.toml: line 7: using '[tool.sip.metadata]' to specify the project metadata is deprecated and will be removed in SIP v7.0.0, use '[project]' instead
>       Traceback (most recent call last):
>         File "/home/steve/cfclient/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
>           main()
>         File "/home/steve/cfclient/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
>           json_out['return_val'] = hook(**hook_input['kwargs'])
>                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>         File "/home/steve/cfclient/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 152, in prepare_metadata_for_build_wheel
>           whl_basename = backend.build_wheel(metadata_directory, config_settings)
>                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>         File "/tmp/pip-build-env-r5z_avhr/overlay/lib/python3.11/site-packages/sipbuild/api.py", line 46, in build_wheel
>           project = AbstractProject.bootstrap('wheel',
>                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>         File "/tmp/pip-build-env-r5z_avhr/overlay/lib/python3.11/site-packages/sipbuild/abstract_project.py", line 92, in bootstrap
>           project.setup(pyproject, tool, tool_description)
>         File "/tmp/pip-build-env-r5z_avhr/overlay/lib/python3.11/site-packages/sipbuild/project.py", line 587, in setup
>           self.apply_user_defaults(tool)
>         File "/tmp/pip-install-6_s_kbyq/pyqt6_f99b94393aac49b8b8a5684d5c65ab35/project.py", line 66, in apply_user_defaults
>           super().apply_user_defaults(tool)
>         File "/tmp/pip-build-env-r5z_avhr/overlay/lib/python3.11/site-packages/pyqtbuild/project.py", line 70, in apply_user_defaults
>           super().apply_user_defaults(tool)
>         File "/tmp/pip-build-env-r5z_avhr/overlay/lib/python3.11/site-packages/sipbuild/project.py", line 237, in apply_user_defaults
>           self.builder.apply_user_defaults(tool)
>         File "/tmp/pip-build-env-r5z_avhr/overlay/lib/python3.11/site-packages/pyqtbuild/builder.py", line 69, in apply_user_defaults
>           raise PyProjectOptionException('qmake',
>       sipbuild.pyproject.PyProjectOptionException
>       [end of output]
>   
>   note: This error originates from a subprocess, and is likely not a problem with pip.
> error: metadata-generation-failed
> 
> × Encountered error while generating package metadata.
> ╰─> See above for output.
> 
> note: This is an issue with the package mentioned above, not pip.
> hint: See above for details.
> 
```

There seems to be a problem with a a deprecated feature in SIP(?) but I've been unable to tell it to go ahead regardless.

Just in case there was an oddity, I've tried this on a Pi 5 with the 32-bit Raspberry Pi OS, a Pi 4, a Pinebook Pro running Manjaro and running Debian.  I also tried it _without_ the `pip install --upgrade pip`. The process I'm using does work on my Intel desktop, but not on any ARM core machine I've tried.

Happy to try any other suggestions.

Cheers,
Steve

**Reply by ataffanel** [reply]
I could finaly test and I am seeing the same problems as you. The main problem is that the PyQT wheel (prebuilt binary) is not built for ARM64 .... Which is very unfortunate since pyqt is very hard to compile.

I see 2 ways forward:
 - It might work with the Debian-provided pyqt. Its verison is a bit older but it might work.
 - We can switch to PySide6 which is distributed as a binary wheel for arm64 . I have tried a bit and it requires a lot of small changes everywhere in the client, but we would be open to switch to pyside6 if that worked for raspi.

I will not have much more time to works more on it right now so if anyone wants to give it a go I created a ticket in the client: https://github.com/bitcraze/crazyflie-clients-python/issues/698.
